plugins/uast-kotlin/src/org/jetbrains/uast/kotlin/declarations/KotlinUVariable.kt

```diff
@@ -110,18 +153,33 @@ class KotlinUVariable(
     override fun getContainingFile(): PsiFile {
         return super.getContainingFile()
     }
+
 }
 
 open class KotlinUParameter(
         psi: PsiParameter,
-        override val sourcePsi: KtElement?,
+        final override val sourcePsi: KtElement?,
         givenParent: UElement?
 ) : AbstractKotlinUVariable(givenParent), UParameter, PsiParameter by psi {
 
-    override val javaPsi = unwrap<UParameter, PsiParameter>(psi)
+    final override val javaPsi = unwrap<UParameter, PsiParameter>(psi)
 
     override val psi = javaPsi
 
+    private val isLightConstructorParam by lz { psi.getParentOfType<PsiMethod>(true)?.isConstructor }
+
+    private val isKtConstructorParam by lz { sourcePsi?.getParentOfType<KtCallableDeclaration>(true)?.let { it is KtPrimaryConstructor } }
+
+    override fun acceptsAnnotationTarget(target: AnnotationUseSiteTarget?): Boolean {
+        if (sourcePsi !is KtParameter) return false
+        if (isKtConstructorParam == isLightConstructorParam && target == null) return true
+        when (target) {
+            AnnotationUseSiteTarget.CONSTRUCTOR_PARAMETER -> return isLightConstructorParam == true
+            AnnotationUseSiteTarget.SETTER_PARAMETER -> return isLightConstructorParam != true
+            else -> return false
```

Comment:
Looks like this won't accept `@param:Foo` annotation on regular (non-constructor) parameters.

Reply:
Also please don't forget about the `@receiver` target; it also maps to parameters.

Reply:
But can `@param:Foo` be used anywhere else except primary constructor? for `@receiver` fixed

Reply:
Yes, you can specify the use site target anywhere even if it's redundant.

Reply:
> '@param:' annotations could be applied only to primary constructor parameters

Idea says to me